### PR TITLE
Add TypeScript Definitions

### PR DESCRIPTION
**Pull request by mortend**
This adds a Definitions file for TypeScript, helpful for code-completion and as documentation.

This file is automatically picked up by TypeScript editors when Fuselibs is installed to `node_modules/` via `npm install` or `npm link`. Tested in Visual Studio Code.

This PR contains:
- [ ] Changelog
- [ ] Documentation
- [ ] Tests

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,2151 @@
...
+                    },
+                    body: JSON.stringify({
+                        registration_ids: [regID],
+                        data: {
+                            notification: {
+                                alert: {
+                                    title: 'Well would ya look at that!',
+                                    body: 'Hello from some other app'
+                                }
+
```
Comment (kusma) [inline index.d.ts]: Most other methods have documentation, why doesn't this?

Comment (kusma) [inline index.d.ts]: Most other methods have documentation, why doesn't this?

[the remark this answers is not in the document]
Comment (kusma) [reply index.d.ts]: Right, I now read what the reason is... But I think we should rename the method instead (but keep a compatibility shim around for backwards-compatibility)...

**Comment by kusma** [comment]
> The biggest blocker to fully automate the generation of this, is that the source dataset is hosted on a different repository. (See [mortend/docs@f55cd5d](https://github.com/mortend/docs/commit/f55cd5dd133a7f59145914c152b76cd056399683).)

Can't we check out the docs repo on CI somewhere, and reach it from there?

Anyway, the whole concept of our documentation (both dataset and generator) being scattered between the docs-repo and the fuselibs/uno repos, is a major source of pain when doing anything docs-related. I have a vague plan to try to fix this: merge the docs generator into uno, and make a proper "uno doc" command that builds html directly from source code + md-files. Then we can generate docs for each uno and fuselibs release on CI, and upload these to github releases as well as deploying them somewhere for hosting. And to get there, I also need to move all docs into each repo

It would mean that our docs becomes split between uno and fuselibs, but I think that's OK...

File: index.d.ts (hunk 1 above)
Comment (kusma) [reply index.d.ts]: I'm not sure I understand. We have a comment saying "Copy a video to the camera roll." for the CopyVideoToCameraRoll method. Isn't that enough?

Comment (mortend) [reply index.d.ts]: The former doc-comment only contain doc-attributes and no other text, so the documentation string (found in api-docs JSON files) is empty/missing, which caused my generator to output this method without having documentation.

Comment (mortend) [reply index.d.ts]: It looks like that comment is missing the `@scriptmethod` attribute which is required to be picked up by my generator.

**Comment by mortend** [comment]
What you're describing sounds like a better solution, though involving more work than I had in mind for this project. When we get a proper "uno doc" command, it will probably be helpful in generating this file in a better way. I guess I was looking for a quicker way to add TypeScript support.

I currently work with local-built fuselibs most of the time, and checking out the docs repo on CI will not work when using `npm link` on local-builds. Locally, I don't think we want to check out the docs repo, build C# code, apply patches etc, during build.

File: index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mortend) [reply index.d.ts]: I can look into renaming those methods.

**Comment by kusma** [comment]
> I currently work with local-built fuselibs most of the time, and checking out the docs repo on CI will not work when using `npm link` on local-builds. Locally, I don't think we want to check out the docs repo, build C# code, apply patches etc, during build.

Right, good point.

**Comment by mortend** [comment]
Rebased and applied changes from #1262. Also dropped the commit that comments out the `delete` methods.

**Comment by mortend** [comment]
If we don't want to add this file to this git repo, I suppose another alternative is to publish it as a separate NPM package, `fusejs-typescript` or something...

**Comment by mortend** [comment]
Whoops, I just made a blunder while merging `release-1.11` into `master`... I have fixed `master` now, but it looks like GitHub now thinks this PR has been merged... Sorry about this! :/

**Comment by mortend** [comment]
Regarding TypeScript Definitions, I now think I'll go for the separate package approach and create a new `fusejs-types` repo for this.

`fusejs-types` can live its own life independent from `fuselibs`, and more easily be updated on demand. It'll be less of a problem if Definitions get out of sync with current documentation, and we can put the generator that made this file in the same repo (instead of in `docs`).

Maybe in the future, when we have better uno doc-tooling, we'll be better equipped to keep things in sync/automated, and can move the Definitions file back to `fuselibs`. Thank you for the feedback.

**Comment by kusma** [comment]
I don't think it's a huge deal, it's just some extra maintenance and risk of accidentally stale documentation. We can merge it to master if it's worth it.

**Comment by mortend** [comment]
In general, I think it's better to distribute definitions together with their implementations, so I feel at `fuselibs` is where they ultimately belong.

Maybe I'll open a new PR with this again. It looks like this PR is now locked forever and can't be reopened.

**Pull request merged**